### ISSUE TITLE
AO keeps overriding and making Dolly stand when not collapsed

### ISSUE THREAD

**Issue by emmamei** [emmamei/cdkey#29]
When kneeling using FS AO, the Key keeps overriding the AO and making the dolly stand.

**Comment by silkymesmeriser** [comment]
For some reason this issue wouldn't replicate for me when I tried to test it however I am not sure why because I can see the flaw in the code causing it and it should have affected me also.

The issue was in my new animation code, as well as the part I already mentioned with the non blocking code it also has an automatic refresh.  The intention was that it would only kill any key animations that shouldn't be running when keyAnimation was set to empty string however I forgot to take into account the issue that would come up from the aoChange(1) line.

New code does an ongoing refresh when the doll is meant to be animated by the key (ie collapsed at the moment but will be use for poses also), then it does a single cycle running through all the animations in it's inventory and does llStopAnimation on each just in case when we do add poses back someone going through the menus too fast somehow manages to get several running at once, then it unsets a variable so that it stops doing anything else until keyAnimation changes to something non empty.

I believe this should fix the issue but would appreciate verification on that.

**Comment by emmamei** [comment]
The problem is now the inverse: it has become impossible to deactivate AO when the key is active. It also still deactivates the AO briefly after a little period of time.

The Key should **never** affect anims unless collapsing and uncollapsing - or posing.

**Comment by silkymesmeriser** [comment]
I have rewritten this function again in my latest code and it should have fixed this albeit accidentally lol

My code I am working on at the moment has a couple of regressions in other areas I need to fix before committing it but try replacing the section of ifPermissions() in Main.lsl which deals with PERMISSION_TRIGGER_ANIMATION with this:

```
    if (perm & PERMISSION_TRIGGER_ANIMATION && isAttached) {
        if (keyAnimation != "") {
            aoControl(0);

            list animList; integer i; integer animCount; key animKey = llGetInventoryKey(keyAnimation);
            while ((animList = llGetAnimationList(dollID)) != [ animKey ]) {
                animCount = llGetListLength(animList);
                for (i = 0; i < animCount; i++) {
                    if (llList2Key(animList, i) != animKey) llStopAnimation(llList2Key(animList, i));
                }
                llStartAnimation(keyAnimation);
            }
        } else if (keyAnimation == "" && clearAnim) {
            list animList = llGetAnimationList(dollID); 
            integer i; integer animCount = llGetInventoryNumber(20);
            for (i = 0; i < animCount; i++) {
                key animKey = llGetInventoryKey(llGetInventoryName(20, i));
                if (llListFindList(animList, [ llGetInventoryKey(llGetInventoryName(20, i)) ]) != -1)
                    llStopAnimation(animKey);
            }
            clearAnim = 0;
            aoControl(1);
        }
    }
```

**Comment by emmamei** [comment]
There's two problems I see right off the bat - but will make it work: firstly, there is no variable `isAttached` nor is there one called `clearAnim`. The current first two `if` statements does the same as the one here.

I've added the code and updated the repository to match.

**Comment by silkymesmeriser** [comment]
isAttached is in my version I forgot that it wasn't committed yet.  Also the extra aoControl(1) outside the if you don't want that is what forces the AO to on even when not posed etc.

All this messing with the animation code has had a positive side effect though I just finally got around to re-implementing poses on my end, it is a surprisingly small amount of code but then it is able to share with the collapsing code for the animation and share with the code I wrote for afk and collapsed mode to prevent movement while posed etc.
